File: backend/app/schemas/order.py
# backend/app/schemas/order.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import datetime
from decimal import Decimal


class OrderItemCreate(BaseModel):
    """Schema cho item trong đơn hàng"""
    book_id: str = Field(..., description="ID của sách")
    quantity: int = Field(..., gt=0, description="Số lượng")
    
    model_config = ConfigDict(from_attributes=True)


class OrderCreate(BaseModel):
    """Schema để tạo đơn hàng mới"""
    user_id: Optional[str] = None  # Sẽ được gán từ current_user
    shipping_address: str = Field(..., min_length=5, description="Địa chỉ giao hàng")
    payment_method_id: str = Field(..., description="ID phương thức thanh toán")
    voucher_code: Optional[str] = Field(None, description="Mã giảm giá (nếu có)")
    items: List[OrderItemCreate] = Field(..., min_length=1, description="Danh sách sách trong đơn")
    
    model_config = ConfigDict(from_attributes=True)


class BookInOrderDetail(BaseModel):
    """Thông tin sách trong order detail"""
    book_id: str
    title: str
    cover_image_url: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)


class OrderDetailResponse(BaseModel):
    """Schema cho chi tiết item trong đơn hàng"""
    detail_id: int
    book_id: str
    quantity: int
    unit_price: Decimal
    book: Optional[BookInOrderDetail] = None
    
    model_config = ConfigDict(from_attributes=True)


class OrderResponse(BaseModel):
    """Schema response chi tiết đơn hàng"""
    order_id: str
    user_id: str
    total_amount: Decimal
    status_id: str
    order_status: str
    shipping_address: str
    payment_method_id: str
    payment_method_name: Optional[str] = None
    created_at: datetime
    order_details: List[OrderDetailResponse] = []
    
    model_config = ConfigDict(from_attributes=True)


class UserOrderHistoryResponse(BaseModel):
    """Schema cho lịch sử đơn hàng của user"""
    total: int
    orders: List[OrderResponse]
    
    model_config = ConfigDict(from_attributes=True)


class OrderStatusUpdate(BaseModel):
    """Schema để cập nhật trạng thái đơn hàng"""
    new_status: str = Field(..., description="Trạng thái mới: confirmed, shipping, completed, cancelled")
    
    model_config = ConfigDict(from_attributes=True)


class OrderCancelResponse(BaseModel):
    """Schema response khi hủy đơn"""
    message: str
    order_id: str
    order_status: str
    
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/order.ts
// frontend/src/api/order.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://127.0.0.1:8000';

export interface OrderItem {
  book_id: string;
  quantity: number;
}

export interface OrderCreate {
  shipping_address: string;
  payment_method_id: string;
  voucher_code?: string;
  items: OrderItem[];
}

export interface BookInOrder {
  book_id: string;
  title: string;
  cover_image_url: string | null;
}

export interface OrderDetailResponse {
  detail_id: number;
  book_id: string;
  quantity: number;
  unit_price: number;
  book: BookInOrder | null;
}

export interface OrderResponse {
  order_id: string;
  user_id: string;
  total_amount: number;
  order_status: string;
  status_id: string;
  shipping_address: string;
  payment_method_id: string;
  payment_method_name: string | null;
  created_at: string;
  order_details: OrderDetailResponse[];
}

export interface UserOrderHistoryResponse {
  total: number;
  orders: OrderResponse[];
}

// ✅ FIX: Thay thế any bằng kiểu dữ liệu cụ thể
export interface AdminOrderSummary {
  order_id: string;
  user_id: string;
  user_name: string;
  total_amount: number;
  order_status: string;
  created_at: string;
}

export interface AdminOrderListResponse {
  total: number;
  orders: AdminOrderSummary[];
}

export interface OrderStatusUpdateResponse {
  message: string;
  order_id: string;
  new_status: string;
}

// Helper để lấy token
function getAuthToken(): string | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem('access_token');
}

// Helper để tạo headers với auth
function getAuthHeaders(): HeadersInit {
  const token = getAuthToken();
  return {
    'Content-Type': 'application/json',
    ...(token ? { 'Authorization': `Bearer ${token}` } : {})
  };
}

/**
 * Tạo đơn hàng mới
 */
export async function createOrder(orderData: OrderCreate): Promise<OrderResponse> {
  console.log("📤 FRONTEND SENDING ORDER:", JSON.stringify(orderData, null, 2));
  
  try {
    const response = await fetch(`${API_BASE_URL}/api/orders/`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify(orderData),
    });

    const responseData = await response.json();
    console.log("📥 BACKEND RESPONSE:", JSON.stringify(responseData, null, 2));

    if (!response.ok) {
      throw new Error(
        typeof responseData.detail === 'string' 
          ? responseData.detail 
          : 'Tạo đơn hàng thất bại'
      );
    }

    return responseData as OrderResponse;
    
  } catch (error) {
    console.error("❌ FETCH ERROR:", error);
    throw error;
  }
}

/**
 * Lấy lịch sử đơn hàng của user hiện tại
 */
export async function getMyOrders(
  skip: number = 0,
  limit: number = 20,
  statusFilter?: string
): Promise<UserOrderHistoryResponse> {
  const params = new URLSearchParams();
  params.append('skip', skip.toString());
  params.append('limit', limit.toString());
  if (statusFilter) params.append('status_filter', statusFilter);

  const url = `${API_BASE_URL}/api/orders/my-orders?${params.toString()}`;

  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || 'Lấy danh sách đơn hàng thất bại');
    }

    const data = await response.json();
    return data as UserOrderHistoryResponse;
  } catch (error) {
    console.error("❌ Fetch error:", error);
    throw error;
  }
}

/**
 * Lấy chi tiết đơn hàng
 */
export async function getOrderDetail(orderId: string): Promise<OrderResponse> {
  const response = await fetch(`${API_BASE_URL}/api/orders/${orderId}`, {
    method: 'GET',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || 'Lấy chi tiết đơn hàng thất bại');
  }

  return response.json() as Promise<OrderResponse>;
}

/**
 * Hủy đơn hàng - CHỈ khi đơn đang Chờ xử lý
 */
export async function cancelOrder(orderId: string): Promise<OrderStatusUpdateResponse> {
  const response = await fetch(`${API_BASE_URL}/api/orders/${orderId}/cancel`, {
    method: 'PUT',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || 'Hủy đơn hàng thất bại');
  }

  return response.json() as Promise<OrderStatusUpdateResponse>;
}

/**
 * User xác nhận đã nhận hàng
 */
export async function confirmDelivery(orderId: string): Promise<OrderStatusUpdateResponse> {
  const response = await fetch(`${API_BASE_URL}/api/orders/${orderId}/confirm-delivery`, {
    method: 'PUT',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || 'Xác nhận nhận hàng thất bại');
  }

  return response.json() as Promise<OrderStatusUpdateResponse>;
}

/**
 * Admin: Cập nhật trạng thái đơn hàng
 */
export async function updateOrderStatus(
  orderId: string,
  newStatus: string
): Promise<OrderStatusUpdateResponse> {
  const response = await fetch(
    `${API_BASE_URL}/api/orders/admin/${orderId}/status?new_status=${newStatus}`,
    {
      method: 'PUT',
      headers: getAuthHeaders(),
    }
  );

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || 'Cập nhật trạng thái thất bại');
  }

  return response.json() as Promise<OrderStatusUpdateResponse>;
}

/**
 * Admin: Lấy tất cả đơn hàng
 */
export async function getAllOrdersAdmin(
  skip: number = 0,
  limit: number = 20,
  statusFilter?: string
): Promise<AdminOrderListResponse> {
  const params = new URLSearchParams();
  params.append('skip', skip.toString());
  params.append('limit', limit.toString());
  if (statusFilter) params.append('status_filter', statusFilter);

  const url = `${API_BASE_URL}/api/orders/admin/all?${params.toString()}`;

  const response = await fetch(url, {
    method: 'GET',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || 'Lấy danh sách đơn hàng thất bại');
  }

  return response.json() as Promise<AdminOrderListResponse>;
}